docs/concepts.md

```diff
@@ -67,6 +68,118 @@ that are specific to a particular Node (like per-process UNIX file descriptors),
+- the **integrity** check ensures that the value sent is trusted by the
+  destination
+
+More concretely, if `L_a = (S_a, I_a)` (where `S_a` and `I_a` are the secrecy
+and integrity components respectively), and `L_b = (S_b, I_b)` then `L_a ⊑ L_b`
+iff `S_a ⊆ S_b` and `I_a ⊇ I_b`. Notably, the integrity part of the check is
+"flipped" by convention (see "Integrity Considerations for Secure Computer
+Systems" below). More fundamental than adhering to convention, by flipping the
+integrity order in this way, we can retain the intuitive meaning behind
+integrity tags as representing trust by a principal (trust by a user, trust on a
+subject, etc.).
+
+Intuitively, data can only flow from `a` to `b` if:
+
+- `b` is **at least as secret** as `a`
+- `a` is **at least as trusted** as `b`
+
+The least privileged label is "public_trusted" (also referred to as `⊥`,
+pronounced "bottom"), which corresponds to a node or channel which has only
+observed public data, and its inputs are not endorsed with any level of
+integrity; in this label, both secrecy and integrity components are empty sets.
+
+For instance, if integrity of `a` is `{s_0, s_1}`, and integrity of `a` is
+`{i_0, i_1}`, then:
+
+- data can flow from `a` to `b` if `b`'s secrecy is `{s_0, s_1, s_2}`, since `b`
+  is "at least as secret" than `a`
+- data cannot flow from `a` to `b` if `b`'s secrecy is `{s_0}`, since `b` is
+  "less secret" than `a` (in particular, `b` is not \$SOMETHING by `s_1`)
+- data cannot flow from `a` to `b` if `b`'s integrity is `{i_0, i_1, i_2}`,
+  since `a` is "less trusted" than `b` (in particular, `a` is not trusted by
+  `i_2`)
+- data can flow from `a` to `b` if `b`'s integrity is `{i_0}`, since `a` is "at
+  least as trusted" than `b`
+
+In particular:
+
+- A node with label `L_w` may write to a channel with label `L_c` iff
+  `L_w ⊑ L_c`.
+- A node with label `L_r` may read from a channel with label `L_c` iff
+  `L_c ⊑ L_r`.
+
+If a node tries to write to or read from a channel that it is not allowed to,
+based on the `⊑` relation, the operation fails immediately, and no side effects
+are performed.
+
+It follows that bi-directional communication between nodes `a` and `b` is only
+allowed (via two uni-directional channels) if
+`(L_a ⊑ L_b) ∧ (L_b ⊑ L_a) ⇒ L_a = L_b`, i.e. if `a` and `b` have identical
+secrecy and integrity.
+
+More details on Information Flow Control may be found in the following
+references:
+
+- [Information Flow Control for Standard OS Abstractions](https://pdos.csail.mit.edu/papers/flume-sosp07.pdf)
+- [Flow-Limited Authorization](https://www.cs.cornell.edu/andru/papers/flam/flam-csf15.pdf)
+- [Integrity Considerations for Secure Computer Systems](http://seclab.cs.ucdavis.edu/projects/history/papers/biba75.pdf)
+
+### gRPC and user labels
+
+For incoming gRPC connections, each gRPC invocation is represented by a message
+containing two channels:
+
+- a "request receiver" channel whose label has:
```

Comment:
nit: Node doesn't get a channel, it gets a handle to the read half of a channel.

Reply:
Done.